torch/nn/modules/module.py

```diff
@@ -193,13 +195,44 @@ def _apply(self, fn):
         for module in self.children():
             module._apply(fn)
 
-        for param in self._parameters.values():
+        for key in self._parameters.keys():
+            # We subtract the refcount by 2 here, because there are two references to
+            # the `self._parameters[key]` tensor that we expect to always exist:
+            #
+            # 1. The reference held by `self._parameters` to its values.
+            # 2. The temporary reference `obj` as an argument to the `sys.getrefcount(obj)` function call
+            #    (for more details, see: https://docs.python.org/3/library/sys.html#sys.getrefcount).
+            param_extra_refcount = sys.getrefcount(self._parameters[key]) - 2
+
+            # We create `param` here after obtaining the refcount of `self._parameters[key]`, instead
+            # of using `for key, param in self._parameters.items():`, because the latter approach would
+            # unnecessarily bump the refcount of the `self._parameters[key]` tensor, which is undesirable.
+            param = self._parameters[key]
             if param is not None:
                 # Tensors stored in modules are graph leaves, and we don't
                 # want to create copy nodes, so we have to unpack the data.
-                param.data = fn(param.data)
-                if param._grad is not None:
-                    param._grad.data = fn(param._grad.data)
+                param_applied = fn(param)
+                if param._is_same_impl_type(param_applied):
+                    if param.device != param_applied.device:
+                        if param_extra_refcount > 0:
+                            warnings.warn("Attempted to change device of `{}`, but there are \
+{} extra reference(s) to that tensor which would be broken by the device change, in future releases \
+of PyTorch.".format(key, param_extra_refcount))
+                    self._parameters[key].data = param_applied
```

Comment:
can we please break up this PR?  This is trying to do a bunch of different things:
1) Fix moving to XLA
2) Do proper version tracking of module parameters (sometimes)
3) Warn about future breaking changes (without first introducing correct APIs).